src/client/components/views/editView/leftSidebar/dropCallback.js

```diff
@@ -1,32 +1,73 @@
+import { Point } from '@projectstorm/geometry';
+import { PNLClasses } from '../../../../../constants';
 
 export function onNodeDrop(monitor, node, engine) {
-
+  const options = new Map();
+  const offset = monitor?.getClientOffset();
+  const height = 619;
+  const width = 266;
+  const position = { x: offset.x - width / 4, y: offset.y - height / 4 };
   pnlStore.dispatch(updateMechanismCount());
   const currentCount = pnlStore.getState().general.mechanismCount;
-  const name = `${node.type}${currentCount}`;
-
-  // Get the client offset (mouse coordinates)
-  const clientOffset = monitor.getClientOffset();
-
-  // Convert the client coordinates to the engine's relative coordinates
-  const position = engine.getRelativeMousePoint({
-    clientX: clientOffset.x,
-    clientY: clientOffset.y,
-  });
-
-  // Set the position as an extra property
-  let extra = {
-    position: {
-      x: position.x,
-      y: position.y
+  const name = `${node.name} ${currentCount}`;
+  const nodeType = node.type.toUpperCase();
+
+  if (engine) {
+    const model = engine.getModel();
+    const selectedNodes = engine.getModel().getSelectedEntities();
+
+    if (node.type === PNLClasses.PROJECTION) {
+      if (selectedNodes.length !== 2) return;
+
+      options.set('id', name);
+      options.set('name', name);
+      options.set('variant', PNLClasses[nodeType]);
+      options.set('shape', PNLClasses[nodeType]);
+      options.set('color', PNLClasses[nodeType]);
+      const sourcePort = selectedNodes[0].getPort(
+        selectedNodes[0].getOptions().ports[1].getName()
+      );
+      const targetPort = selectedNodes[1].getPort(
+        selectedNodes[1].getOptions().ports[0].getName()
+      );
+
+      const link = new MetaLinkModel(Object.fromEntries(options));
+      link.setSourcePort(sourcePort);
+      link.setTargetPort(targetPort);
+
+      model.addLink(link);
+      engine.repaintCanvas();
+    } else {
+      options.set('id', name);
+      options.set('name', name);
+      options.set('variant', 'node-blue');
+      options.set('width', width);
+      options.set('height', height);
+      options.set('selected', false);
+      options.set('pnlClass', PNLClasses[nodeType]);
+      options.set('shape', PNLClasses[nodeType]);
+      options.set('graphPath', [null]);
+      options.set('depth', 0);
+      options.set('ports', [
+        new MetaPort('in', 'in', PortTypes.INPUT_PORT, undefined, undefined),
+        new MetaPort('out', 'out', PortTypes.OUTPUT_PORT, undefined, undefined),
+      ]);
+      options.set('position', new Point(position.x, position.y));
+      options.set('localPosition', new Point(position.x, position.y));
+
+      const newNode = new MetaNodeModel(Object.fromEntries(options));
+      return model.addNode(newNode);
     }
-  };
 
-  const newNode = NodeFactory.createNode(node.type, name, extra);
-  const newNodeModel = newNode.getMetaNode().toModel()
-  ModelSingleton.getInstance().getMetaGraph().addNode(newNodeModel)
```

Comment:
Also please use [this api](https://github.com/MetaCell/PsyNeuLinkView/blob/metacell/src/client/model/graph/MetaGraph.ts#L185) to add new nodes, because it will add the node first to our MetaGraph and consequentially to the react-diagrams model. We need to do that so both models are consistent.